### PR TITLE
jwk #5: jwk observation

### PR DESCRIPTION
**Pull request by zjma**

### CODE REVIEW

File: crates/aptos-jwk-consensus/src/jwk_observer.rs [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+            String::from_utf8(issuer),
+            String::from_utf8(config_url)
+        );
+        Self {
+            close_tx,
+            join_handle,
+        }
+    }
+
+    async fn thread_main(
```
Comment (zekun000) [inline crates/aptos-jwk-consensus/src/jwk_observer.rs:86]: you mean main_loop?

File: crates/aptos-jwk-consensus/src/jwk_observer.rs [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    join_handle: JoinHandle<()>,
+}
+
+impl JWKObserver {
+    pub fn spawn(
+        my_addr: AccountAddress,
+        issuer: Issuer,
+        config_url: Vec<u8>,
+        fetch_interval: Duration,
+        observation_tx: aptos_channel::Sender<(), (Issuer, Vec<JWK>)>,
```
Comment (zekun000) [inline crates/aptos-jwk-consensus/src/jwk_observer.rs:64]: this doesn't need to be aptos channel?

File: crates/aptos-jwk-consensus/src/jwk_observer.rs [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+/// Given an Open ID configuration URL, fetch its JWKs.
+pub async fn fetch_jwks(my_addr: AccountAddress, config_url: Vec<u8>) -> Result<Vec<JWK>> {
+    if cfg!(feature = "smoke-test") {
+        use reqwest::header;
+        let maybe_url = String::from_utf8(config_url);
+        let jwk_url = maybe_url?;
+        let client = reqwest::Client::new();
+        let JWKsResponse { keys } = client
+            .get(jwk_url.as_str())
+            .header(header::COOKIE, my_addr.to_hex())
```
Comment (zekun000) [inline crates/aptos-jwk-consensus/src/jwk_observer.rs:33]: is the difference just the header?

File: crates/aptos-jwk-consensus/src/jwk_observer.rs [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+use futures::{FutureExt, StreamExt};
+use move_core_types::account_address::AccountAddress;
+use serde::{Deserialize, Serialize};
+use std::time::Duration;
+use tokio::{sync::oneshot, task::JoinHandle, time::MissedTickBehavior};
+
+#[derive(Serialize, Deserialize)]
+struct OpenIDConfiguration {
+    issuer: String,
+    jwks_uri: String,
```
Comment (danielxiangzl) [inline crates/aptos-jwk-consensus/src/jwk_observer.rs:16]: uri or url?

File: crates/aptos-jwk-consensus/src/jwk_observer.rs (hunk 3 above)
Comment (zjma) [reply crates/aptos-jwk-consensus/src/jwk_observer.rs:33]: in smoke test it's 1 request; in reality it's 1 request for open_id_config, then a 2nd call for open_id_config.jwk_url

File: crates/aptos-jwk-consensus/src/jwk_observer.rs (hunk 2 above)
Comment (zjma) [reply crates/aptos-jwk-consensus/src/jwk_observer.rs:64]: why not? what's a better option for this?

File: crates/aptos-jwk-consensus/src/jwk_observer.rs [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+    keys: Vec<serde_json::Value>,
+}
+
+/// Given an Open ID configuration URL, fetch its JWKs.
+pub async fn fetch_jwks(my_addr: AccountAddress, config_url: Vec<u8>) -> Result<Vec<JWK>> {
+    if cfg!(feature = "smoke-test") {
+        use reqwest::header;
+        let maybe_url = String::from_utf8(config_url);
+        let jwk_url = maybe_url?;
+        let client = reqwest::Client::new();
```
Comment (danielxiangzl) [inline crates/aptos-jwk-consensus/src/jwk_observer.rs:30]: nit: request

Comment (zjma) [reply crates/aptos-jwk-consensus/src/jwk_observer.rs:30]: it's the name of the library

File: crates/aptos-jwk-consensus/src/jwk_observer.rs (hunk 1 above)
Comment (danielxiangzl) [reply crates/aptos-jwk-consensus/src/jwk_observer.rs:86]: call `start` to be consistent with the codebase

File: crates/aptos-jwk-consensus/src/jwk_observer.rs [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+        close_rx: oneshot::Receiver<()>,
+    ) {
+        let mut interval = tokio::time::interval(fetch_interval);
+        interval.set_missed_tick_behavior(MissedTickBehavior::Delay);
+        let mut close_rx = close_rx.into_stream();
+        loop {
+            tokio::select! {
+                _ = interval.tick().fuse() => {
+                    let result = fetch_jwks(my_addr, open_id_config_url.clone()).await;
+                    debug!("observe_result={:?}", result);
```
Comment (danielxiangzl) [inline crates/aptos-jwk-consensus/src/jwk_observer.rs:101]: print more information? like version, issuer, my_id?

Comment (zjma) [reply crates/aptos-jwk-consensus/src/jwk_observer.rs:101]: will have a new PR to cover log + metrics

**Review by danielxiangzl — APPROVED** [review]

File: crates/aptos-jwk-consensus/src/jwk_observer.rs (hunk 2 above)
Comment (zekun000) [reply crates/aptos-jwk-consensus/src/jwk_observer.rs:64]: just a normal channel? aptos channel is to provide round robin and drop ability

**Review by zekun000 — APPROVED** [review]

File: crates/aptos-jwk-consensus/src/jwk_observer.rs (hunk 4 above)
Comment (zjma) [reply crates/aptos-jwk-consensus/src/jwk_observer.rs:16]: uri by design. that's what's returned [here](https://www.facebook.com/.well-known/openid-configuration/)

**Pull request merged**